File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        
        if( (head.next == null && n == 1)  || head == null) return null;
        
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode p1 = head;
        ListNode prev = head;

        
        for(int i=0; i<n; i++){
            p1 = p1.next;
        }
        
        if(p1 == null){
            return head.next;
        }
        
        while(p1.next != null){   
            p1 = p1.next;
            prev = prev.next;  
        }
        
        prev.next = prev.next.next;
        
        return head;
        
    }
}

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let slow = head;
    let fast = head;
    let prev = null;

    if(slow.next === null) return slow.next;

    let i = 0;
    while( i < n){
        i++;
        fast = fast.next;
    }

    while(fast){
        prev = slow;
        slow = slow.next;
        fast = fast.next;
    }

    if(prev){
        prev.next = slow.next;
    }else{
        head = head.next;
    }
    

    return head;
};